### PR TITLE
WI #1860 Add semantic support for environment mnemonics

### LINKED ISSUE #1860: Semantic error: Symbol xxx is not referenced for mnemonic-name defined in special names
**Describe the bug**
Programs with statements WREITE AFTER ADVANCING mnemonic-name1 compiles with the above error even if the mnemonic-name is defined in special-names clause.

To Reproduce, compile a program like this:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID.     TCOFM117.
       ENVIRONMENT DIVISION.
       CONFIGURATION SECTION.
       SOURCE-COMPUTER. IBM-370.
       OBJECT-COMPUTER. IBM-370.
       SPECIAL-NAMES.
      * Syntax is:
      *  Environment-name IS Mnemonic-name
      *  - Valid values of Environment-name are listed in IBM
      *    specifications
      *  - Mnemonic-name is an user defined word
      
      
      *    can only be used in ACCEPT
           SYSIN  is SYSIN-M
      *Ko define same environment-name twice
      *A duplicate"SYSIN"clause was found.  Scanning was resumed at the
      *next area"A"item or the start of the next clause.
           SYSIN  is SYSIN-M2
           SYSIPT    SYSIPT-M
      
      *    can only be used in DISPLAY
           SYSOUT   is SYSOUT-M
           SYSLIST     SYSLIST-M
           SYSLST   is SYSLST-M
           SYSPUNCH is SYSPUNCH-M
           SYSPCH   is SYSPCH-M
      
      *    Ok in ACCEPT and DISPLAY
           CONSOLE is CONSOLE-M
      
      *    Ok in WRITE ADVANCING
           C01 is C01-M
           C02    C02-M
           C03 is C03-M
      
      *    KO syntax only for UPSI switch
           C04 is C04-M
             ON STATUS IS C04-M-ON
             OFF STATUS IS C04-M-OFF
      
      *    KO syntax only for UPSI switch
           C05
             ON STATUS IS C05-ON
             OFF STATUS IS C05-OFF
      
      
           CSP    IS CSP-M
           S01    IS S01-M
           S02       S02-M
           S03      S03-M
           S04    IS S04-M
      *Ko reuse same mnemonic-name
      *Mnemonic-name"S04-M"was previously defined.  The duplicate
      *environment-name clause was discarded.
           S05    IS S04-M
           AFP-5A IS AFP-5A-M
      
      
           decimal-point is comma.
       INPUT-OUTPUT SECTION.
       FILE-CONTROL.
               SELECT  FSYS020  ASSIGN TO UT-S-SYS020.
       DATA DIVISION.
       FILE SECTION.
       FD  FSYS020
           LABEL RECORD STANDARD
           RECORDING MODE F
           BLOCK 0 RECORDS
           DATA RECORD E-EDIT.
       01  E-EDIT                 PIC X(150).
      *-----------------------*
       WORKING-STORAGE SECTION.
       01 Var1 pic X(08).
      *Ko The name"S03-M"was used for an item that was
      *   not defined as a dat  References to this name
      *   may be resolved incorrectly.
       01 S03-M pic X.
      
       PROCEDURE DIVISION.
      
      
      *-ACCEPT--------------------
      *OK usage of environment-name
           accept Var1 from SYSIN
           accept Var1 from SYSIPT
           accept Var1 from CONSOLE
      
      *OK usage of mnemonic-name
           accept Var1 from SYSIN-M
           accept Var1 from SYSIPT-M
           accept Var1 from CONSOLE-M
      
      *-DISPLAY-------------------
      *OK usage of environment-name
           display " " upon  SYSOUT
           display " " upon  SYSLIST
           display " " upon  SYSLST
           display " " upon  SYSPUNCH
           display " " upon  SYSPCH
           display " " upon  CONSOLE
      
      *OK usage of mnemonic-name
           display " " upon  SYSOUT-M
           display " " upon  SYSLIST-M
           display " " upon  SYSLST-M
           display " " upon  SYSPUNCH-M
           display " " upon  SYSPCH-M
           display " " upon  CONSOLE-M
      
      
      *KO mnemonic-name cannot be used as variables
      *    "SYSIN"was not defined as a data-name.
      *    The statement was discarded.
           display SYSIN
           display SYSIPT
           display CONSOLE
           display SYSOUT
           display SYSLIST
           display SYSLST
           display SYSPUNCH
           display SYSPCH
           display CONSOLE
      
      *KO mnemonic-name cannot be used as variables
      *    "SYSIN-M"was not defined as a data-name.
      *    The statement was discarded.
           display SYSIN-M
           display SYSIPT-M
           display CONSOLE-M
           display SYSOUT-M
           display SYSLIST-M
           display SYSLST-M
           display SYSPUNCH-M
           display SYSPCH-M
           display CONSOLE-M
      
      *    Presence of Open before write is not currently checked
      *    by our parser
           OPEN OUTPUT FSYS020
           write E-EDIT after advancing C01-M
           write E-EDIT after advancing C02-M
           write E-EDIT after advancing C03-M
      
           write E-EDIT after advancing S01-M
           write E-EDIT after advancing S02-M
      *    KO "S03-M"was not a uniquely defined name.
      *      The definition to be used could not be determined
      *      from the context.  The reference to the name was discarded.
      *      Ambiguous reference to S03-M
           write E-EDIT after advancing S03-M
      *    Warning Ambiguous reference to S03-M
           write E-EDIT after advancing S04-M
      *    KO "S05-M"was not defined as a data-name.
      *      The statement was discarded.
      *      Unable to resolve reference to S05-M
           write E-EDIT after advancing S05-M
           CLOSE FSYS020
      
           goback
           .
      
       end program TCOFM117.
```

Original sample of the issue
```cobol
        IDENTIFICATION DIVISION.                          
       PROGRAM-ID.    TESTC01.                           
       ENVIRONMENT DIVISION.                             
       CONFIGURATION SECTION.
       SPECIAL-NAMES.      C01 IS TO-TOP-OF-PAGE.
       INPUT-OUTPUT SECTION.                             
       FILE-CONTROL.                                     
           SELECT REPOUT                                 
                  ASSIGN TO UT-S-REPORT.                 
       DATA DIVISION.                                    
       FILE SECTION.                                     
       FD  REPOUT                                        
               RECORD CONTAINS 120 CHARACTERS            
               LABEL RECORDS ARE OMITTED                 
               DATA RECORD IS REPREC.                    
       01  REPREC                     PIC X(120).        
       WORKING-STORAGE SECTION.                          
       01  REPHDR1  PIC X(120)                           
                   VALUE '-----------------------------'.
       PROCEDURE DIVISION.                               
       PROG-START.                                       
                WRITE REPREC FROM REPHDR1                
                   AFTER ADVANCING TO-TOP-OF-PAGE.       
           GOBACK.                                       
       END PROGRAM TESTC01.                              
```

Expected behavior:
Mnemonic-Name represented in Symbol-Table and no compile error.

**Technical current situation**
- `SymbolType.MnemonicForEnvironmentName` and `SymbolType.EnvironmentName`  already exists.
- `Accept` and `Display` statements allow an `environment-name` or an `mnemonic-name`. Their SymbolReference are ambiguous and can be of 2 types: `SymbolType.MnemonicForEnvironmentName` and `SymbolType.EnvironmentName` 
- `Write` statement only allow `mnemonic-name`. Its `SymbolReference` is not ambiguous and can only be of type `SymbolType.MnemonicForEnvironmentName`
- Grammar only allow `environment-name` or an `mnemonic-name` at the same place. E.g. in the grammar a `SymbolReference` cannot be of type `mnemonic-name` and `DataName`.
- The list of `environment-name` is fixed for a statement.

To analyse :
- Do we put the fixed list of allowed `environment-name` directly in the grammar ?
   - It means that `SymbolReference` of type `SymbolType.EnvironmentName` will be resolved directly by the grammar. The `SymbolReference` for both `SymbolType.EnvironmentName` and `SymbolType.MnemonicForEnvironmentName` will NOT be ambiguous.
   - No other check of this `SymbolType`. 
   - It's optimized with incremental mode
 - OR do we keep the ambiguity and we resolve these Symbol later : during `CodeElement` step to optimize incremental mode? During `CrossChecker` to keep all symbols resolution together ?
- Create a new entry in `SymbolTable` to hold `mnemonic`. I think we can store directlty `SymbolDefinition` or even `string` of mnemonic.
- In `CrossChecker` check is `SymbolReference` is or can be of Type `MnemonicForEnvironmentName`.

**How to tests**
Standard program tests.

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #1860

### CODE REVIEW

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 1]
```diff
@@ -649,6 +649,12 @@ public override bool Visit(DataDefinition dataDefinition)
             DataDefinitionEntry dataDefinitionEntry = dataDefinition.CodeElement;
             if (dataDefinitionEntry == null) return true;
 
+            if (dataDefinition.SymbolTable.IsEnvironmentMnemonicNameDefined(dataDefinition.Name))
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: done

File: TypeCobol/Compiler/CodeModel/SymbolTable.cs [hunk 2]
```diff
@@ -1211,6 +1211,75 @@ public IEnumerable<Program> GetPrograms()
 
         #endregion
 
+        #region SPECIAL NAMES
+
+        private readonly IDictionary<string, List<SymbolDefinition>> EnvironmentMnemonics =
+            new Dictionary<string, List<SymbolDefinition>>(StringComparer.OrdinalIgnoreCase);
+
+        private SymbolTable GetRootGlobalTable()
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/CodeModel/SymbolTable.cs]: I've renamed the first method as suggested. However the second one has already 23  usages, it existed prior to this PR.

File: TypeCobol/Compiler/CodeModel/SymbolTable.cs [hunk 3]
```diff
@@ -1219,9 +1288,10 @@ public IEnumerable<Program> GetPrograms()
         /// <typeparam name="T"></typeparam>
         /// <param name="table"></param>
         /// <param name="symbol"></param>
-        private static void Add<T>([NotNull] IDictionary<string, List<T>> table, [NotNull] T symbol) where T : Node
+        /// <param name="getName"></param>
+        private static void Add<T>([NotNull] IDictionary<string, List<T>> table, [NotNull] T symbol, [NotNull] Func<T, string> getName)
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/CodeModel/SymbolTable.cs]: ok

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs [hunk 4]
```diff
@@ -414,10 +417,22 @@ public override void EnterSpecialNamesParagraph(CodeElementsParser.SpecialNamesP
...
+					{
+						// Duplicate environment, add to duplicate list and skip definition
+						duplicateEnvironments.Add(environmentNameContext.environmentName());
+						continue;
+					}
+					
+					var mnemonicForEnvironmentName = _cobolWordsBuilder.CreateMnemonicForEnvironmentNameDefinition(environmentNameContext.mnemonicForEnvironmentNameDefinition());
+					if (paragraph.MnemonicsForEnvironmentNames.ContainsKey(mnemonicForEnvironmentName))
+					{
+						// Duplicate mnemonic, add tu duplicate list and skip definition
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: done

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 5]
```diff
@@ -882,7 +888,122 @@ public override bool Visit(Exec exec)
...
+                    if (dataDefinition == null)
+                    {
+                        //This is a mnemonic, nothing to do.
+                    }
+                    else
+                    {
+                        //This is still ambiguous...
+                        DiagnosticUtils.AddError(write, $"Ambiguous reference to '{referenceToResolve.Name}', the definition to be used could not be determined from the context.", referenceToResolve, MessageCode.SemanticTCErrorInParser);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: yes it's shorter and still clear

**Comment by fm-117** [comment]
I have to add some test cases for the visibility of mnemonics in case of nested pgms.
More commits to come...

File: TypeCobol.Test/Parser/Programs/Cobol85/MnemonicsForEnvironment_Nested3.rdzPGM.txt [hunk 6]
```diff
@@ -0,0 +1,127 @@
+--- Diagnostics ---
+Line 57[8,31] <27, Error, Syntax> - Syntax error : The name 'SYSIN-M' was used for an item that was not defined as a data-name. References to this name may be resolved incorrectly.
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Test/Parser/Programs/Cobol85/MnemonicsForEnvironment_Nested3.rdzPGM.txt]: Message updated

**Review by smedilol — APPROVED** [review]

**Pull request merged**